instrumentation/opentelemetry-instrumentation-pika/src/opentelemetry/instrumentation/pika/pika_instrumentor.py

```diff
@@ -0,0 +1,119 @@
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from logging import getLogger
+from typing import Any, Callable, Collection, Dict, Optional
+
+import wrapt
+from pika.adapters import BlockingConnection
+from pika.channel import Channel
+
+from opentelemetry import trace
+from opentelemetry.instrumentation.instrumentor import BaseInstrumentor
+from opentelemetry.instrumentation.pika import utils
+from opentelemetry.instrumentation.pika.package import _instruments
+from opentelemetry.instrumentation.pika.version import __version__
+from opentelemetry.instrumentation.utils import unwrap
+from opentelemetry.trace import Tracer, TracerProvider
+
+_LOG = getLogger(__name__)
+_CTX_KEY = "__otel_task_span"
+
+_FUNCTIONS_TO_UNINSTRUMENT = ["basic_publish"]
+
+
+class PikaInstrumentor(BaseInstrumentor):  # type: ignore
+    # pylint: disable=attribute-defined-outside-init
+    @staticmethod
+    def _instrument_consumers(
+        consumers_dict: Dict[str, Callable[..., Any]], tracer: Tracer
+    ) -> Any:
+        for key, callback in consumers_dict.items():
+            decorated_callback = utils._decorate_callback(
+                callback, tracer, key
+            )
+            setattr(decorated_callback, "_original_callback", callback)
+            consumers_dict[key] = decorated_callback
+
+    @staticmethod
+    def _instrument_basic_publish(channel: Channel, tracer: Tracer) -> None:
+        original_function = getattr(channel, "basic_publish")
+        decorated_function = utils._decorate_basic_publish(
+            original_function, channel, tracer
+        )
+        setattr(decorated_function, "_original_function", original_function)
+        channel.__setattr__("basic_publish", decorated_function)
+        channel.basic_publish = decorated_function
+
+    @staticmethod
+    def _instrument_channel_functions(
+        channel: Channel, tracer: Tracer
+    ) -> None:
+        if hasattr(channel, "basic_publish"):
+            PikaInstrumentor._instrument_basic_publish(channel, tracer)
+
+    @staticmethod
+    def _uninstrument_channel_functions(channel: Channel) -> None:
+        for function_name in _FUNCTIONS_TO_UNINSTRUMENT:
+            if not hasattr(channel, function_name):
+                continue
+            function = getattr(channel, function_name)
+            if hasattr(function, "_original_function"):
+                channel.__setattr__(function_name, function._original_function)
+
+    @staticmethod
+    def instrument_channel(
+        channel: Channel, tracer_provider: Optional[TracerProvider] = None,
+    ) -> None:
+        tracer = trace.get_tracer(__name__, __version__, tracer_provider)
```

Comment:
Please use this [flag](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/instrumentation/opentelemetry-instrumentation-flask/src/opentelemetry/instrumentation/flask/__init__.py#L270) to avoid double instrumentation.

Reply:
Nice! Done